gateway/throttler/throttler.go

```diff
@@ -0,0 +1,134 @@
+package throttler
+
+//go:generate mockgen -destination=../../mocks/gateway/throttler.go -package=mocks_gateway github.com/rudderlabs/rudder-server/gateway/throttler GetThrottler
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/internal/throttling"
+	"github.com/rudderlabs/rudder-server/services/stats"
+)
+
+const (
+	throttlingAlgoTypeGCRA = "gcra"
+)
+
+type Limiter interface {
+	// Allow returns true if the limit is not exceeded, false otherwise.
+	Allow(ctx context.Context, cost, rate, window int64, key string) (bool, func(context.Context) error, error)
+}
+
+type GetThrottler interface {
+	CheckLimitReached(workspaceId string) (bool, error)
+}
+
+type Factory struct {
+	Stats        stats.Stats
+	limiter      Limiter
+	throttlers   map[string]*Throttler // map key is the workspaceId
+	throttlersMu sync.Mutex
+}
+
+// New constructs a new Throttler Factory
+func New(stats stats.Stats) (*Factory, error) {
+	f := Factory{
+		Stats:      stats,
+		throttlers: make(map[string]*Throttler),
+	}
+	if err := f.initThrottlerFactory(); err != nil {
+		return nil, err
+	}
+	return &f, nil
+}
+
+func (f *Factory) CheckLimitReached(workspaceId string) (bool, error) {
+	t := f.get(workspaceId)
+	return t.checkLimitReached(workspaceId)
+}
+
+func (f *Factory) get(workspaceId string) *Throttler {
+	f.throttlersMu.Lock()
+	defer f.throttlersMu.Unlock()
+	if t, ok := f.throttlers[workspaceId]; ok {
+		return t
+	}
+
+	var conf throttlingConfig
+	conf.readThrottlingConfig(workspaceId)
+	f.throttlers[workspaceId] = &Throttler{
+		limiter: f.limiter,
+		config:  conf,
+	}
+	return f.throttlers[workspaceId]
+}
+
+func (f *Factory) initThrottlerFactory() error {
+	throttlingAlgorithm := config.GetString("Gateway.throttler.algorithm", throttlingAlgoTypeGCRA)
+
+	var (
+		err  error
+		l    *throttling.Limiter
+		opts []throttling.Option
+	)
+	if f.Stats != nil {
+		opts = append(opts, throttling.WithStatsCollector(f.Stats))
+	}
+	switch throttlingAlgorithm {
+	case throttlingAlgoTypeGCRA:
+		l, err = throttling.New(append(opts, throttling.WithInMemoryGCRA(0))...)
+	default:
+		return fmt.Errorf("invalid throttling algorithm: %s", throttlingAlgorithm)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create throttler: %w", err)
+	}
+
+	f.limiter = l
+
+	return nil
+}
+
+type Throttler struct {
+	limiter Limiter
+	config  throttlingConfig
+}
+
+// checkLimitReached returns true if we're not allowed to process the number of event
+func (t *Throttler) checkLimitReached(key string) (limited bool, retErr error) {
+	ctx := context.TODO()
```

Comment:
[minor] since this a newly created method, no reason not to expose the interface with a context
```suggestion
func (t *Throttler) checkLimitReached(ctx context.Context, key string) (limited bool, retErr error) {
```